### PR TITLE
PRfD - List Pull Requests UI

### LINKED ISSUE #8173: PRfD - WebUI - List PRs page
Create `Pull Requests` tab in the lakeFS WebUI,
And it'll show the list of PRs available for this repo.

This includes:
- Handling PRfD UI routing.
- Getting all the PRs from the API.
- Displaying all the PRs in the tab.
- Handle pagination.
- Add Open/Closed tabs.
- Add "Create PR" button.

### PR DESCRIPTION
**Pull request by itaigilo**
Closes #8173.

## Change Description

Creating UI for Listing Pull Requests for a repo.
Including the new `Pull Requests` tab.

-----
 
Screenshot 2024-09-17 at 20 32 09

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -342,6 +342,9 @@ $(UI_DIR)/node_modules:
 gen-ui: $(UI_DIR)/node_modules  ## Build UI web app
 	cd $(UI_DIR) && $(NPM) run build
 
+gen-ui-watch: $(UI_DIR)/node_modules  ## Build UI web app and watch for changes
```
Comment (itaigilo) [inline Makefile:345]: Helpful for UI dev.

File: webui/src/lib/api/index.js [hunk 2]
```diff
@@ -567,6 +567,62 @@ class Tags {
...
+    async list(repoId, state = "open", prefix = "", after = "", amount = DEFAULT_LISTING_AMOUNT) {
+        // const query = qs({prefix, after, amount});
+        // const response = await apiRequest(`/repositories/${encodeURIComponent(repoId)}/pulls?` + query);
+        // if (response.status !== 200) {
+        //     throw new Error(`could not list pulls: ${await extractError(response)}`);
+        // }
+        // return response.json();
+
+        // TODO: this is for development purposes only
+        console.log("list pulls", {repoId, state, prefix, after, amount})
```
Comment (itaigilo) [inline webui/src/lib/api/index.js:580]: As written - this is temp, until the API will be ready.

File: webui/src/lib/components/repository/changes.jsx [hunk 3]
```diff
@@ -153,7 +153,7 @@ export const ChangesTreeContainer = ({results, delimiter, uriNavigator,
                     <div>{changesTreeMessage}</div>
                     <Card>
                         <Card.Header>
-                            <span className="float-start">
+                            <span className="float-start w-100">
```
Comment (itaigilo) [inline webui/src/lib/components/repository/changes.jsx:156]: For some reason, `width: 100%` was added globally to `float-start`, which is wrong.
Removed it, and added `w-100` to all the places using it (and not overriding).

File: webui/src/lib/components/repository/tabs.jsx [hunk 4]
```diff
@@ -59,6 +60,11 @@ export const RepositoryNavTabs = ({ active }) => {
             <Link active={active === 'tags'} href={`/repositories/${repoId}/tags`} component={NavItem}>
                 <TagIcon/> Tags
             </Link>
+            {showPulls &&
+                <Link active={active === 'pulls'} href={`/repositories/${repoId}/pulls`} component={NavItem}>
+                    <GitCompareIcon/> Pull Requests
```
Comment (itaigilo) [inline webui/src/lib/components/repository/tabs.jsx:65]: Using the same icon as `Compare`. We'll need to change one of them.

File: webui/src/lib/components/repository/tabs.jsx [hunk 5]
```diff
@@ -59,6 +60,11 @@ export const RepositoryNavTabs = ({ active }) => {
             <Link active={active === 'tags'} href={`/repositories/${repoId}/tags`} component={NavItem}>
                 <TagIcon/> Tags
             </Link>
+            {showPulls &&
+                <Link active={active === 'pulls'} href={`/repositories/${repoId}/pulls`} component={NavItem}>
```
Comment (itaigilo) [inline webui/src/lib/components/repository/tabs.jsx:64]: Note - I've used `pulls` (and not `pull-requests`) pretty much everywhere.

File: webui/src/pages/repositories/repository/pulls/pulls.jsx
Comment (itaigilo) [inline webui/src/pages/repositories/repository/pulls/pulls.jsx:1]: This is based on the `Tags` page - 
I changed mainly what I needed, and left the rest very similar.

File: webui/src/pages/repositories/repository/pulls/pulls.jsx [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+    const authorLink =
+        <Link href={{
+            // TODO: to where this link should lead?
+            pathname: '/repositories/:repoId/user/:userId',
+            params: {repoId: repo.id, userId: pull.author}
+        }}>
+            {pull.author}
+        </Link>;
+    return (
+        <ListGroup.Item className="pull-row">
```
Comment (N-o-Z) [inline webui/src/pages/repositories/repository/pulls/pulls.jsx:30]: Look at the example of commits tab, it should look the same IMO.
Font size, alignment hyperlink marking and the visuals (images and colors) should have the same style.
This tab should look and feel like the other tabs

File: webui/src/pages/repositories/repository/pulls/pulls.jsx [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+const PullStatus = {
+    open: "open",
+    closed: "closed",
+    merged: "merged",
+}
+
+const PullsList = ({repo, after, prefix, onPaginate}) => {
+    const router = useRouter()
+    const [refresh, setRefresh] = useState(true);
+    // TODO: pullState should be persistent in the url and saved as a url param?
```
Comment (N-o-Z) [inline webui/src/pages/repositories/repository/pulls/pulls.jsx:65]: state is a query param in GitHub

**Review by N-o-Z — CHANGES_REQUESTED** [review]
Really nice work! And great small and contained PR.
I'm blocking because I believe the tab design is not aligned with the overall lakeFS UI design and needs adjusting.
Since this is a major part of this PR I'm requesting changes

**Review by itaigilo — COMMENTED** [review]
@N-o-Z Thanks for the review.

Your comments have been addressed, PTAL again.

File: webui/src/lib/components/repository/tabs.jsx [hunk 8]
```diff
@@ -1,15 +1,15 @@
...
 
 import {useRefs} from "../../hooks/repo";
 import {Link, NavItem} from "../nav";
 import {useRouter} from "../../hooks/router";
 import {RefTypeBranch} from "../../../constants";
 
-// TODO: this is temp, until PRfD will be ready
-const showPulls = false;
+// TODO (gilo): this is temp, until PRfD will be ready
+const showPulls = true;
```
Comment (N-o-Z) [inline webui/src/lib/components/repository/tabs.jsx:12]: Need to revert this to `false` 😃

**Review by N-o-Z — APPROVED** [review]
I'm approving with 2 concerns:
1. Need to revert the flag!
2. The images still need adjusting - either in this PR or open a new issue for it. But we should have the same visuals as in compare:
image

**Comment by itaigilo** [comment]
@N-o-Z thanks for spotting the flag - reverted.

Regarding how the source/dest looks like: 
I generally agree - I honestly don't have a clear understanding on how this should look from the PRD,
So I think that the best way to approach this will be to schedule a UI review once all the functionality is there.
I'll schedule such review, and will add an Issue for it.

**Pull request merged**